numpy/core/src/multiarray/nditer_pywrap.c

```diff
@@ -1272,13 +1275,45 @@ npyiter_copy(NewNpyArrayIterObject *self)
 
     iter->started = self->started;
     iter->finished = self->finished;
+    iter->open = self->open;
 
     return (PyObject *)iter;
 }
 
+/*
+ * Resolve writeback semantics and make operands read-only
+ */
+static PyObject *
+npyiter_close(NewNpyArrayIterObject *self)
+{
+    npy_intp iop, nop;
+    PyArrayObject **operands;
+
+    if (self->iter == NULL) {
+        PyErr_SetString(PyExc_ValueError,
+                "Iterator is invalid");
+        return NULL;
+    }
+
+    nop = NpyIter_GetNOp(self->iter);
+    operands = self->operands;
+
+    for (iop=0; iop<nop; iop++) {
+        if (PyArray_ResolveWritebackIfCopy(operands[iop]) > 0) {
+            PyArray_CLEARFLAGS(operands[iop], NPY_ARRAY_WRITEABLE);
+        }
+    }
+    self->open = 0;
+    Py_RETURN_NONE;
+}
+
 static PyObject *
 npyiter_iternext(NewNpyArrayIterObject *self)
 {
+    if (self->open == 0) {
+        PyErr_SetString(PyExc_RuntimeError, "Iterator closed");
```

Comment:
This one (and only this one) should raise `StopIteration` for consistency with `generator` objects

Either that, or `__next__` should be different from `iternext`, and the former should raise StopIteration

Reply:
Do you mean ``npyiter_next`` which is assigned to the ``tp_iternext`` slot? I am confused who calls ``npyiter_iternext`` which is exposed as a ``nditer.iternext`` method, it does not seem to be documented in the python typeobject model for [python2](https://docs.python.org/2/c-api/typeobj.html) nor [python3](https://docs.python.org/3/c-api/typeobj.html)

Reply:
if I iterate over a closed file object, it raises a ``ValueError``, not a ``StopIteration`` error

Reply:
You're right, I do mean `npyiter_next`. `.iternext()` is part of our API, not a python one.

Good counterexample with `file`